Synchronize WebGL contexts with page rendering

jdm opened the pull request:
The code that deals with a integrating a shared GL context with the display list that Webrender receives doesn't work right now. Webrender has an API that allows locking the context, but the period between the end of layout (when script starts running again) and when WR actually processes the display list means that the GL context can be modified so that what WR sees is not what was supposed to be composited. This change marks GL contexts for WebGL canvases as locked as soon as they are processed by layout, causing any future WebGL commands that affect the to context to be buffered and replayed once WebRender unlocks the GL context.

---
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #21838
- [x] These changes do not require tests because it's nondeterministic behaviour.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21841)
<!-- Reviewable:end -->

jdm commented:
r? @emilio 
Let me know if this is more than you want to deal with and I can find somebody else.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9995)

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -183,6 +212,13 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
 
+    /// Handles a request to prevent a GL context from being modified until further notice.
+    fn handle_prevent_modification(&mut self, context_id: WebGLContextId) {
+        if !self.buffered_contexts.contains_key(&context_id) {
```
nit: This can use `entry` instead to avoid two hash table lookups.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -115,6 +120,27 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
+        let target_id = match msg {
+            WebGLMsg::CreateContext(..) |
+            WebGLMsg::Lock(..) |
+            WebGLMsg::Unlock(..) |
```
We should probably assert that we don't receive lock messages while it's unlocked, or similar. Otherwise the code below in `handle_unlock` won't work. Though I think this is a pre-existing issue.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -200,13 +236,23 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
+        {
+            let data = Self::make_current_if_needed(context_id, &self.contexts, &mut self.bound_context_id)
+                            .expect("WebGLContext not found in a WebGLMsg::Unlock message");
+            let info = self.cached_context_info.get_mut(&context_id).unwrap();
```
nit: while you're touching this, using the indexing operator may give a nicer error message if it fails.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -64,6 +65,9 @@ pub struct WebGLThread<VR: WebVRRenderHandler + 'static> {
     /// Texture ids and sizes used in DOM to texture outputs.
     dom_outputs: FnvHashMap<webrender_api::PipelineId, DOMToTextureData>,
+    /// Tracking GL contexts that must not be modified but can still be
+    /// targeted by WebGL commands.
```
This should document that this map only has keys for contexts that are currently locked, _and_ for which a `PreventModification` message has been sent.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1336,7 +1338,11 @@ impl LayoutThread {
+            guards.clone(),
+            true,
+            data.reflow_goal.needs_display(),
+            &map);
```
nit: paren on the next line.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1592,7 +1598,13 @@ impl LayoutThread {
+                guards,
+                false,
+                goal.needs_display(),
+                &snapshots
```
nit: I think rustfmt would add a trailing comma here.

emilio commented on components/canvas/webgl_thread.rs:
```diff
@@ -183,6 +212,13 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
     }
 
+    /// Handles a request to prevent a GL context from being modified until further notice.
+    fn handle_prevent_modification(&mut self, context_id: WebGLContextId) {
```
Should this assert that the context is locked? Otherwise there's no guarantee that the context would be unlocked, right?

emilio reviewed (commented):
np, I think I can review this :)

Some comment below, I think it looks reasonable, though I think I need to re-page the locking / unlocking scheme is supposed to work to see if this is sound. We can probably take advantage that we're going to see us in person on Monday to double-check this :)

Thanks!

emilio commented:
Another thing to check is whether WebRender is guaranteed to give us an unlock() message. I'd expect it to not do so if the image is culled (because the element is off-screen for example).

MortimerGoro commented:
As an optimization we could start the msg buffering only when a Draw operation arrives (e.g. glDrawElements, glDrawArrays or glClear). It should be safe to run other methods.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21543) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
This runs into problems where JS will block on compositing. I'm going to try using textures like Gecko instead of a single offscreen framebuffer so we can just tell webrender "this is the frame; draw it" and keep working on a separate texture.

MortimerGoro commented:
That's a good idea. We are also going to need render to texture (SurfaceTexture) for WebVR  integration with FxR.

jdm committed:
Use local offscreen_gl_context.

jdm committed:
webgl: Swap the current draw buffer every time a webgl canvas is marked dirty.

Refactors the WebGL thread and canvas rendering context to reflect the fact that
WebRender image keys are immutable. Any readback that is necessary occurs right
before the draw buffer is swapped.

jdm committed:
Remove offscreen_gl_context dependency from canvas_traits and script.

jdm commented:
This is definitely a huge improvement in debug builds, at least.

jdm commented:
@bors-servo retry
* #22262

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 3c229f1df44e104ad96414918c1445e7d6b3e863 with merge 40dec74f30cdcf433b880bff72c2cfe88b07a538...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/10883)

jdm commented:
Current theory about the pixel 2 brokenness:
1. WR locks the external image and fetches the texture id for the complete frame
1. the canvas thread processes the lock message and returns the current texture id for the complete frame
1. the canvas thread receives a message to present the current frame and swaps the textures, resetting the previously complete texture to black
1. WR draws the texture id that has now been reset to black

More specifically, the locking mechanism currently implemented is broken by the switch to use multiple textures. The swapping code should only run when the texture is unlocked; otherwise we should just reset the current frame and skip displaying it, I suppose.

jdm commented:
The flickering is not helped by dealing with the issue described in the previous comment. It _is_ helped by commenting out the call to reset_draw_buffer_contents in https://github.com/servo/rust-offscreen-rendering-context/pull/131.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #22385) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
No longer necessary with the introduction of #24482!
